elements/research/ddiotune.cc

```diff
@@ -0,0 +1,385 @@
+	sprintf(msr_file,"/dev/cpu/%d/msr",cpu);
+
+	fd = open(msr_file, O_WRONLY);
+	if (fd < 0)
+		return fd;
+
+	ret = pwrite(fd, val, sizeof(uint64_t), msr);
+
+	close(fd);
+
+	return ret;
+}
+
+/* Initialize PCI access data structure */
+void
+DDIOTune::init_pci_access(void)
+{
+	pacc = pci_alloc();           /* Get the pci_access structure */
+	pci_init(pacc);               /* Initialize the PCI library */
+	pci_scan_bus(pacc);           /* We want to get the list of devices */
+}
+
+/* Find the responsible PCI root for the input NIC */
+struct pci_dev*
+DDIOTune::find_ddio_device(uint8_t nic_bus)
+{
+	struct pci_dev* dev;
+	for(dev = pacc->devices; dev; dev=dev->next) {
+		pci_fill_info(dev,PCI_FILL_IDENT | PCI_FILL_BASES | PCI_FILL_NUMA_NODE | PCI_FILL_PHYS_SLOT);
+		/* 
+		 * Find the proper PCIe root based on the nic device
+		 * For instance, if the NIC is located on 0000:17:00.0 (i.e., BDF)
+		 * 0x17 is the nic_bus (B)
+		 * 0x00 is the nic_device (D)
+		 * 0x0	is the nic_function (F)
+		 */
+		if(/*dev->func == 0 && dev->dev == 0  && */pci_read_byte(dev,PCI_SUBORDINATE_BUS) == nic_bus ) {
+		return dev;
+		}
+	}
+	printf("Could not find the proper PCIe root!\n");
+	return NULL;
+}
+
+
+/* Print cache information */
+void
+DDIOTune::print_cache_info()
+{
+	click_chatter("Cache Hierarchy Info");
+	click_chatter("====================");
+	for(int i=0;i<CACHE_LEVELS;i++){
+		click_chatter("Level %d:",i);
+		click_chatter("#Ways: %d",ways[i]);
+		//click_chatter("#Partitions: %"PRIu32,partitions[i]);
+		//click_chatter("Cache Line Size: %d",linesize[i]);
+		click_chatter("#Sets: %d",sets[i]);
+		click_chatter("Size: %d B",size[i]);
+		click_chatter("====================");
+	}
+
+}
+
+/* Check DDIO status (i.e., enabled or disabled) for the input NIC */
+int
+DDIOTune::ddio_status(uint8_t nic_bus)
+{
+	uint32_t val;
+	if(!pacc)
+		init_pci_access();
+
+	struct pci_dev* dev=find_ddio_device(nic_bus);
+	if(!dev){
+		printf("No device found!\n");
+		printf("No device found!\n");
+		exit(1);
+	}
+	/*
+	 * perfctrlsts_0
+	 * bit 3: NoSnoopOpWrEn -> Should be 1b
+	 * bit 7: Use_Allocating_Flow_Wr -> Should be 0b
+	 * Check p. 68 of Intel® Xeon® Processor Scalable Family
+	 * Datasheet, Volume Two: Registers
+	 * May 2019
+	 * link: https://www.intel.com/content/www/us/en/processors/xeon/scalable/xeon-scalable-datasheet-vol-2.html
+	 * Haswell link: https://www.intel.com/content/dam/www/public/us/en/documents/datasheets/xeon-e5-v2-datasheet-vol-2.pdf
+	 */
+	val=pci_read_long(dev,PERFCTRLSTS_0);
+	printf("perfctrlsts_0 val: 0x%" PRIx32 "\n",val);
+	printf("NoSnoopOpWrEn val: 0x%" PRIx32 "\n",val&nosnoopopwren_MASK);
+	printf("Use_Allocating_Flow_Wr val: 0x%" PRIx32 "\n",val&use_allocating_flow_wr_MASK);
+	if(val&use_allocating_flow_wr_MASK)
+		return 1;
+	else
+		return 0;
+}
+
+/* Enable DDIO for the input NIC */
+void
+DDIOTune::ddio_enable(uint8_t nic_bus)
+{
+	uint32_t val;
+	if(!pacc)
+		init_pci_access();
+
+	if(!ddio_status(nic_bus))
+	{
```

Comment:
This bracket should go at the same line with the if condition

Reply:
Reindent will fix this